_posts/2021-01-23-modeling-ui-state.md

```diff
@@ -0,0 +1,196 @@
+
+- `Unit?` - cardinality = 2 (1 + cardinality(Unit))
+- `Boolean?` - 3 (1 + cardinality(Boolean))
+
+The elements of a nullable type `A?` are `null` + the elements of the original type `A` so the cardinality of a nullable type is 1 + cardinality(original). 
+
+### Enums
+
+Another flavor of sum types in Kotlin are enums:
+
+```kotlin
+enum class Color { RED, YELLOW, GREEN }
+```
+
+The cardinality of `Color` is equal to the number of elements, in this case 3. An alternative representation using a sealed class is:
+
+```kotlin
+sealed class Color
+object Red : Color()
+object Yellow : Color()
+object Green : Color()
+```
+
+and it has the same cardinality - 3.
+
+## How does it matter
+
+Thinking about Types as Sets and their cardinality helps with data modelling to avoid a whole class of bugs related to Invalid Data. Let's say we are modeling a traffic light. The possible colors are red, yellow and green. To represent them in code we could use:
+
+- a `String` where `"red"`, `"yellow"` and `"green"` are the valid options and everything else is invalid data. But then the user types `"rad"` instead of red and we have an issue. Or `"yelow"` or `"RED"`. Should all functions validate the data? Should all functions have tests? The root cause of the issue here is cardinality. A String has cardinality of ∞ while the our problem has 3. There are ∞ - 3 possible invalid values.
+
+- a data class `data class Color(val isRed: Boolean, val isYellow: Boolean, val isGreen: Boolean)` - here red is represented as `Color(true, false, false)`. However this still leaves room for invalid data e.g. `Color(true, true, true)`. Again you would need checks and test to ensure values are valid. The cardinality of the data class `Color` is 8 and it has 8 - 3 = 5 illegal values. It's much better then the `String`, but we can do better.
+
+- an enum - `enum class Color { RED, YELLOW, GREEN }` - this has a cardinality = 3 which is exactly the same as the possible valid values of the problem. Illegal values are now impossible, so there is no need for tests that check data validity.
+
+By modelling the data in a way that rules out illegal values, the resulting code will end up shorter, more clear and easier to test. 
+
+> Make sure the set of possible values in code match the set of valid values of the problem and a lot of problems disappear.
+
+## Exposing State from a ViewModel
+
+The problem: build an app that calls a traffic light endpoint and shows the color of the traffic light (red, yellow or green). During the network show a `ProgressBar`, on success show a `View` with the color and in case of an error show a `TextView` with generic text. There is no retry, exactly one view is show at a time.
```

Comment:
"There is no retry, exactly one view is show at a time." -> Not sure this sentence is necessary. In case it is, it should be "shown" instead of "show".